spec/services/create_order_service_spec.rb

```diff
@@ -0,0 +1,94 @@
+      create_order_service = described_class.new(order_params)
+
+      expect { create_order_service.call }.to change { ActionMailer::Base.deliveries.count }.by 1
+    end
+
+    it 'returns false if there is no available online code for reward' do
+      employee = create(:employee, earned_points: 1)
+      reward_with_online_delivery = create(:reward)
+      order_params = { employee_id: employee.id, reward_id: reward_with_online_delivery.id }
+      create_order_service = described_class.new(order_params)
+
+      expect(create_order_service.call).to be false
+      expect(create_order_service.errors.to_s).to include 'Reward is not available now.'
+    end
+  end
+
+  context 'when buying a reward with post delivery method' do
+    it 'returns true, adds new order and adress to database when all params are properly set and does not send email' do
+      employee = create(:employee, earned_points: 1)
+      reward_with_post_delivery = create(:reward, delivery_method: 'post_delivery')
+      order_params = { employee_id: employee.id, reward_id: reward_with_post_delivery.id }
+      address_params = attributes_for(:address)
+      create_order_service = described_class.new(order_params.merge(address_params))
+
+      expect(Order.count).to eq 0
+      expect(create_order_service.call).to be true
+      expect(Order.count).to eq 1
+      expect(ActionMailer::Base.deliveries.count).to eq 0
+      expect(Order.last.address_snapshot.street).to eq address_params[:street]
+      expect(Order.last.address_snapshot.postcode).to eq address_params[:postcode]
+      expect(Order.last.address_snapshot.city).to eq address_params[:city]
+    end
+
+    it "updates employee's address when address already exists" do
+      employee = create(:employee, earned_points: 1)
+      reward_with_post_delivery = create(:reward, delivery_method: 'post_delivery')
+      first_address = create(:address, employee: employee)
+      order_params = { employee_id: employee.id, reward_id: reward_with_post_delivery.id }
+      address_params = attributes_for(:address)
+      create_order_service = described_class.new(order_params.merge(address_params))
+
+      expect(Address.count).to eq 1
+      expect(employee.address.street).to eq first_address.street
+      expect(employee.address.postcode).to eq first_address.postcode
+      expect(employee.address.city).to eq first_address.city
+
+      expect { create_order_service.call }.not_to change(Address, :count)
+
+      employee.reload
+      expect(employee.address.street).not_to eq first_address.street
+      expect(employee.address.postcode).not_to eq first_address.postcode
+      expect(employee.address.city).not_to eq first_address.city
+      expect(employee.address.street).to eq address_params[:street]
+      expect(employee.address.postcode).to eq address_params[:postcode]
+      expect(employee.address.city).to eq address_params[:city]
```

Comment:
if you explicitly define the values of `address_params` in the beginning of this scenario, then you can just test for those values in here. For sure you don't have to test for them being different than something and then for being equal to smth else.

Tests become much more readable when there are explicit values involved - even if they are set to "foo" or 7

so this block may be then changed to something like
```ruby
expect(employee.address).to have_attributes(
  street: 'Downing Street',
  postcode: 'AB00111',
  city: 'London'
)
```

Reply:
Thanks for explanation 👍 Of course I used your suggestion and changed code :)